djtriggers/tests/test_models.py

```diff
@@ -0,0 +1,106 @@
+from datetime import timedelta
+from logging import WARNING
+
+import mock
+import pytest
+from django.test import override_settings
+from django.test.testcases import TestCase
+from django.utils import timezone
+from locking.models import NonBlockingLock
+
+from djtriggers.exceptions import ProcessLaterError, AlreadyProcessedError
+from djtriggers.loggers.base import TriggerLogger
+from djtriggers.models import Trigger
+from djtriggers.tests.factories.triggers import NormalDummyTriggerFactory, RaisesExceptionDummyTriggerFactory
+
+
+class TriggerTest(TestCase):
+    """
+    Due to some bug in celery.contrib.methods.task_method testing the task method needs be done in the following way:
+        Class.TaskMethod(class_instance)
+    """
+
+    def test_handle_execution_success(self):
+        trigger = NormalDummyTriggerFactory()
+        trigger._handle_execution_success()
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
+
+    @mock.patch('django_statsd.clients.statsd')
+    def test_handle_execution_success_use_statsd(self, mock_statsd):
+        trigger = NormalDummyTriggerFactory(process_after=timezone.now())
+        trigger._handle_execution_success(use_statsd=True)
+
+        assert trigger.date_processed
+        assert timezone.now() - trigger.date_processed < timedelta(minutes=1)
+        mock_statsd.incr.assert_called_with(
+            'triggers.{trigger_type}.processed'.format(trigger_type=trigger.trigger_type))
+        mock_statsd.timing.assert_called_with(
+            'triggers.{trigger_type}.process_delay_seconds'.format(trigger_type=trigger.trigger_type),
+            (trigger.date_processed - trigger.process_after).total_seconds())
+
+    def test_handle_execution_failure(self):
+        trigger = NormalDummyTriggerFactory()
+        original_tries = trigger.number_of_tries
+        trigger._handle_execution_failure(Exception())
+
+        assert trigger.number_of_tries == original_tries + 1
+
+    @override_settings(DJTRIGGERS_TRIES_BEFORE_WARNING=3)
+    @mock.patch.object(TriggerLogger, 'log_message')
+    def test_handle_execution_failure_tries_exceeded(self, mock_logger):
```

Comment:
the edge case of the happy path is not tested number_of_tries = 2 should not raise an exception (currently this is not the case)